blockchain/message_pool/src/msgpool/mod.rs

```diff
@@ -398,6 +399,149 @@ pub mod tests {
+        let sender = wallet.generate_addr(SignatureType::Secp256k1).unwrap();
+        let target = wallet.generate_addr(SignatureType::Secp256k1).unwrap();
+        let tma = TestApi::default();
+        tma.set_state_sequence(&sender, 0);
+
+        let (tx, _rx) = flume::bounded(50);
+        let mut services = JoinSet::new();
+        let mpool = MessagePool::new(
+            tma,
+            "mptest".to_string(),
+            tx,
+            Default::default(),
+            Arc::default(),
+            &mut services,
+        )
+        .unwrap();
+        let mut smsg_vec = Vec::new();
+        for i in 0..4 {
+            let msg = create_smsg(&target, &sender, wallet.borrow_mut(), i, 1000000, 1);
+            smsg_vec.push(msg);
+        }
+
+        mpool.add(smsg_vec[0].clone()).unwrap();
+        mpool.add(smsg_vec[1].clone()).unwrap();
+        mpool.add(smsg_vec[2].clone()).unwrap();
+        mpool.add(smsg_vec[3].clone()).unwrap();
+
+        let (p, _) = mpool.pending().unwrap();
+        assert_eq!(p.len(), 4);
+
+        let header_a = mock_block(1, 1);
+
+        mpool
+            .api
+            .inner
+            .lock()
+            .set_block_messages(&header_a, smsg_vec.clone());
+
+        let tipset_a = Tipset::from(&header_a.clone());
+
+        let ts = tipset_a.clone();
+        mpool.api.set_heaviest_tipset(Arc::new(ts));
+
+        // sleep allows for async block to update mpool's cur_tipset
+        tokio::time::sleep(Duration::new(2, 0)).await;
+
+        let (p, _) = mpool.pending().unwrap();
+        assert!(p.is_empty());
+
+        // clear and create a new vector
+        smsg_vec.clear();
+        let sender = wallet.generate_addr(SignatureType::Secp256k1).unwrap();
+
+        for i in 0..4 {
+            let msg = create_smsg(&target, &sender, wallet.borrow_mut(), i, 1000000, 1);
+            smsg_vec.push(msg);
+        }
+
+        mpool.add(smsg_vec[0].clone()).unwrap();
+        mpool.add(smsg_vec[1].clone()).unwrap();
+        mpool.add(smsg_vec[2].clone()).unwrap();
+        mpool.add(smsg_vec[3].clone()).unwrap();
+
+        let (p, _) = mpool.pending().unwrap();
+        assert_eq!(p.len(), 4);
+
+        // build chain a <- b <- c
+        let header_b = mock_block_with_parents(&tipset_a, 1, 1);
+        mpool
+            .api
+            .inner
+            .lock()
+            .set_block_messages(&header_b, smsg_vec.clone());
+        let tipset_b = Tipset::from(&header_b.clone());
+
+        let header_c = mock_block_with_parents(&tipset_b, 1, 1);
+        let tipset_c = Tipset::from(&header_c.clone());
+
+        // Test 1: calling set_heaviest_chain with only head does not clear mpool
+        // messages present in head ancestors
+        let ts_chain = NonEmpty::new(Arc::new(tipset_c.clone()));
+        mpool.api.set_heaviest_chain(ts_chain);
+
+        // sleep allows for async block to update mpool's cur_tipset
+        tokio::time::sleep(Duration::new(2, 0)).await;
```

Comment:
I highly dislike `sleep`s in the tests.

1. It makes them flaky, yielding failures when run on unexpectedly low-end (or busy) workers.
2. It makes tests unnecessarily long. Say the update took 5ms - we are waiting for 2s for nothing.

Can you think of a mechanism that would at least partially mitigate those issues?

Reply:
Yes. I don't like it as well and I'm looking for a way to remove those.

I've put 10ms while waiting for something better.

Reply:
I've added a new method in `Provider` so it's possible for `TestApi` to use `await` instead of `sleep`.